sdk/core/Azure.Core/tests/TestFramework/RecordMatcher.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Azure.Core.Testing
+{
+    public class RecordMatcher
+    {
+        private readonly RecordedTestSanitizer _sanitizer;
+
+        public RecordMatcher(RecordedTestSanitizer sanitizer)
+        {
+            _sanitizer = sanitizer;
+        }
+
+        public HashSet<string> ExcludeHeaders = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "Date",
+            "x-ms-client-request-id",
+            "User-Agent"
+        };
+
+        public virtual int FindMatch(Request request, IList<RecordEntry> entries)
+        {
+            SortedDictionary<string, string[]> headers = new SortedDictionary<string, string[]>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var header in request.Headers)
+            {
+                var gotHeader = request.Headers.TryGetValues(header.Name, out IEnumerable<string> values);
+                Debug.Assert(gotHeader);
+                headers[header.Name] = values.ToArray();
+            }
+
+            _sanitizer.SanitizeHeaders(headers);
+
+            string uri = _sanitizer.SanitizeUri(request.UriBuilder.ToString());
+
+            for (int i = 0; i < entries.Count; i++)
+            {
+                RecordEntry entry = entries[i];
+
+                if (entry.RequestUri == uri &&
+                    entry.RequestMethod == request.Method &&
+                    CompareHeaderDictionaries(headers, entry.RequestHeaders))
```

Comment:
>CompareHeaderDictionaries(headers, entry.RequestHeaders)) [](start = 20, length = 57)

I think that matching on headers might be a bit excessive.  Some headers may be environment specific, but perhaps this means that we should be scrubbing those.  Thoughts?

Reply:
I was trying to start from being as strict as possible and see how it goes, there is a way to opt out headers from comparison.

>  Some headers may be environment specific, but perhaps this means that we should be scrubbing those

My thought was that we are storing the entire connection string so headers should be reproducible if they are not based on scrubbed values but in that case they might need to be scrubbed too.